### DISCUSSION TITLE
Social Commonality Between Repos

### DISCUSSION THREAD

**Discussion by falahat** [datactive/bigbang#516] [Ideas]
One of the things that I'd like to work on for the end of the semester would be to explore the specific people that repositories depend on.

I'd like to start with an Author->Repo graph, where the edge weights represent how many commits that author has made to a given repo.

I'd then like to give each author a "Connector Rating" (we'll find a better name later) that basically represents how much of a role that author plays in connecting all of those repos. However, what I need help figuring out is how to calculate this value.

Some factors that will probably play a role:
- How many repos the author is attached to
- How many commits the author has made to each repo
- How many other authors contribute to the same repositories. Are there 50 people that all commit to Repo A and Repo B or is there only one person that commits to both Repo A and Repo B.

I'd like to draw the author->repo graph but mark the author nodes based on their Connector Rating (make them bigger or something.)

**Comment by sbenthall** [comment]
This is a great idea. I'd like to work with you to flesh this idea out. At the moment, the first thing that comes to mind is the work that @Jack005 was working on last semester. He was looking at betweeness centrality, which is the measure of how often a node lies on the shortest path between all other nodes.

In this case you are looking at a bipartite graph, so you may want to use the bipartite variation on this metric:

http://networkx.lanl.gov/archive/networkx-1.5/reference/generated/networkx.algorithms.bipartite.centrality.betweenness_centrality.html

I don't know if this metric will capture the idea that you are looking for, but it would be an easy first shot at it.

You might also want to look up the idea of "structural holes", as it's related to what you are getting at I think.

http://www.econ.upf.edu/docs/seminars/burt.pdf

Besides looking at existing work in this area, I encourage you to flesh out _why_ this "Connector Ranking" interests you. There are infinite possible mathematical expressions we could compute over this graph. The question is which ones represent meaningful constructs.

If there were some other variable that we were trying to predict in terms of this 'Connector Ranking', then we could compare different ranking metrics in their ability to predict that dependent variable. This is certainly the direction this research is going.

**Comment by sbenthall** [comment]
This paper by Pohl and Diehl (2008) directly pertains to this and should be used as a foundation for this work.

www.st.uni-trier.de/~diehl/pubs/chase08.pdf

**Comment by sbenthall** [comment]
I wonder if this is related to some of the things we've been discussing regarding people's participation in groups, @npdoty

**Comment by npdoty** [comment]
Ah, yes, I have been doing this kind of work with bipartite graphs and seeing the overlaps between participation, but with mailing lists rather than git repositories.

I'm using the bitpartite-specific measure, although I evaluated closeness centrality rather than betweenness centrality in my research notebook:
https://github.com/npdoty/bigbang/blob/ietf-participation/ietf-participation/IETF%20Participants.ipynb
